source/iterator/snapshot.go

```diff
@@ -0,0 +1,166 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iterator
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	sdk "github.com/conduitio/conduit-connector-sdk"
+
+	"github.com/conduitio-labs/conduit-connector-firebolt/client"
+	"github.com/conduitio-labs/conduit-connector-firebolt/source/position"
+)
+
+const (
+	// metadata related.
+	metadataTable = "firebolt.table"
+)
+
+// SnapshotIterator snapshot iterator.
+type SnapshotIterator struct {
+	// firebolt client.
+	client *client.Client
+	// index - current index of element in current batch which iterator converts to record.
+	indexInBatch int
+	// batchID - current batch id, show what batch iterator uses, using in query to get currentBatch.
+	batchID int
+	// batchSize size of batch.
+	batchSize int
+	// currentBatch - rows in current batch from table.
+	currentBatch []map[string]any
+	// name of columns what iterator use for setting key in record.
+	primaryKeys []string
+	// list of columns to reading from table.
+	columns []string
+	// table which iterator read.
+	table string
+}
+
+func NewSnapshotIterator(
+	client *client.Client,
+	batchSize int,
+	table string,
+	columns, primaryKeys []string,
+) *SnapshotIterator {
+	return &SnapshotIterator{
+		client:      client,
+		batchSize:   batchSize,
+		primaryKeys: primaryKeys,
+		columns:     columns,
+		table:       table}
+}
+
+// Setup iterator.
+func (i *SnapshotIterator) Setup(ctx context.Context, p sdk.Position) error {
+	if p != nil {
+		pos, err := position.ParseSDKPosition(p)
+		if err != nil {
+			return nil
+		}
+
+		i.indexInBatch = pos.IndexInBatch + 1
+		i.batchID = pos.BatchID
+	}
+
+	rows, err := i.client.GetRows(ctx, i.table, i.primaryKeys, i.columns, i.batchSize, i.batchID)
+	if err != nil {
+		sdk.Logger(ctx).Debug().Str("table", i.table).Str("primaryKey",
+			strings.Join(i.primaryKeys, ",")).Str("columns", strings.Join(i.columns, ",")).
+			Int("batchSize", i.batchSize).Int("batchID", i.batchID).Msg("get rows parameters")
+
+		return fmt.Errorf("get rows: %w", err)
+	}
+
+	i.currentBatch = rows
+
+	return nil
+}
+
+// HasNext check ability to get next record.
+func (i *SnapshotIterator) HasNext(ctx context.Context) (bool, error) {
+	var err error
+
+	if i.indexInBatch < len(i.currentBatch) {
+		return true, nil
+	}
+
+	if i.indexInBatch >= i.batchSize {
+		i.batchID += i.batchSize
+		i.indexInBatch = 0
+	}
+
+	i.currentBatch, err = i.client.GetRows(ctx, i.table, i.primaryKeys, i.columns, i.batchSize, i.batchID)
+	if err != nil {
+		return false, err
+	}
+
+	if len(i.currentBatch) == 0 || len(i.currentBatch) <= i.indexInBatch {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// Next get new record.
+func (i *SnapshotIterator) Next(ctx context.Context) (sdk.Record, error) {
+	pos := position.NewPosition(i.indexInBatch, i.batchID)
+
+	payload, err := json.Marshal(i.currentBatch[i.indexInBatch])
```

Comment:
Nitpick: maybe it would make the code a bit easier to understand, if `i.currentBatch[i.indexInBatch]` would be extracted as a variable.

Reply:
we can do it but it will be additional memory allocation